File: verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/method/MethodArgumentTypesVerifier.java
package com.jetbrains.pluginverifier.verifiers.method;

import com.intellij.structure.resolvers.Resolver;
import com.jetbrains.pluginverifier.location.ProblemLocation;
import com.jetbrains.pluginverifier.problems.ClassNotFoundProblem;
import com.jetbrains.pluginverifier.utils.VerifierUtil;
import com.jetbrains.pluginverifier.verifiers.VContext;
import org.objectweb.asm.Type;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.MethodNode;

/**
 * @author Sergey Patrikeev
 */
public class MethodArgumentTypesVerifier implements MethodVerifier {
  @Override
  public void verify(ClassNode clazz, MethodNode method, Resolver resolver, VContext ctx) {
    Type methodType = Type.getType(method.desc);
    Type[] argumentTypes = methodType.getArgumentTypes();
    for (Type type : argumentTypes) {
      String argDescr = VerifierUtil.extractClassNameFromDescr(type.getDescriptor());
      if (argDescr == null) continue;
      if (!VerifierUtil.classExistsOrExternal(ctx, resolver, argDescr)) {
        ctx.registerProblem(new ClassNotFoundProblem(argDescr), ProblemLocation.fromMethod(clazz.name, method));
      }
    }


  }
}

File: verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/method/MethodArgumentTypesVerifier.kt
package com.jetbrains.pluginverifier.verifiers.method

import com.jetbrains.pluginverifier.verifiers.VerificationContext
import com.jetbrains.pluginverifier.verifiers.createMethodLocation
import com.jetbrains.pluginverifier.verifiers.extractClassNameFromDescr
import com.jetbrains.pluginverifier.verifiers.resolveClassOrProblem
import org.objectweb.asm.Type
import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.MethodNode

class MethodArgumentTypesVerifier : MethodVerifier {
  override fun verify(clazz: ClassNode, method: MethodNode, ctx: VerificationContext) {
    val methodType = Type.getType(method.desc)
    val argumentTypes = methodType.argumentTypes
    for (type in argumentTypes) {
      val argDescr = type.descriptor.extractClassNameFromDescr() ?: continue
      ctx.resolveClassOrProblem(argDescr, clazz) { createMethodLocation(clazz, method) }
    }


  }
}
